File: cpp-super-tiny-virtual-machine/src/bytecode.cpp
#include "models.h"
bool sortNode(ASTNode *a, ASTNode *b);
Atom *getAtomByVal(Context &kc, string val);
Atom *getAtomByVal(Context &kc, int val);
int bytecodeParse(Context &kc, ASTNode *node, string indent)
{
    auto emit = [&](int index, ASTNode *node)
    {
        emit2(kc.cg, node->type == ASTNodeType::IntLiteral ? OP_TYPE::NUMBER : OP_TYPE::NAME, index);
    };
    switch (node->type)
    {
        int lindex, rindex, index;
    case ASTNodeType::Program:
        for (auto child : node->children)
            bytecodeParse(kc, child, indent);
        break;
    case ASTNodeType::Additive:
    case ASTNodeType::Multiplicative:
    {
        lindex = bytecodeParse(kc, node->children[0], indent);
        rindex = bytecodeParse(kc, node->children[1], indent);
        if (lindex != -1)
            emit(lindex, node->children[0]);
        if (rindex != -1)
            emit(rindex, node->children[1]);

        if (node->text == "+")
            emit1(kc.cg, OP_TYPE::ADD);
        else if (node->text == "-")
            emit1(kc.cg, OP_TYPE::MINUS);
        else if (node->text == "*")
            emit1(kc.cg, OP_TYPE::MUTI);
        else if (node->text == "/")
            emit1(kc.cg, OP_TYPE::DIVID);
        break;
    };
    case ASTNodeType::IntLiteral:
        getAtomByVal(kc, stoi(node->text));
        return kc.atoms.size() - 1;
    case ASTNodeType::Identifier:
        getAtomByVal(kc, node->text);
        return kc.atoms.size() - 1;
    case ASTNodeType::AssignmentStmt:
    case ASTNodeType::IntDeclaration:
    {
        index = bytecodeParse(kc, node->children[0], indent);
        if (index >= 0)
            emit(index, node->children[0]);
        Atom *atom = getAtomByVal(kc, node->text);
        emit2(kc.cg, node->type == ASTNodeType::AssignmentStmt ? OP_TYPE::ASSIGN : OP_TYPE::INT_DECL, kc.atoms.size() - 1);
        break;
    }
    }
    return -1;
}
//可以允许重复的atom，空间换时间O(1)找到index
Atom *getAtomByVal(Context &kc, string val)
{
    Atom *atom = generateAtom(val);
    kc.atoms.push_back(atom);
    return atom;
}
Atom *getAtomByVal(Context &kc, int val)
{
    Atom *atom = generateAtom(val);
    kc.atoms.push_back(atom);
    return atom;
}
bool sortNode(ASTNode *a, ASTNode *b)
{
    return b->children.size() - a->children.size();
}

File: super-tiny-virtual-machine/src/bytecode.ts
import { ASTNode, ASTNodeType, OpType, str2Uint8Array, to_optype_str } from "./models";

/**
 * 遍历AST,生成字节码
 */
const BUFFER_SIZE = 10000;
export default class Bytecode {
    offset: number = 0;
    parse(node: ASTNode, indent: string): ArrayBuffer {
        // console.log("===byte generator===")
        this.offset = 0;
        let opBuffer = new ArrayBuffer(BUFFER_SIZE);//二进制指令buffer
        const dataView = new DataView(opBuffer);
        this.execute(node, indent, dataView);
        return opBuffer.slice(0, this.offset);
    }
    execute(node: ASTNode, indent: string, dataView: DataView): any {
        indent = `${indent}\t`;
        let that = this, value, lvalue, rvalue;
        switch (node.type) {
            case ASTNodeType.Program:
                for (let child of node.children) {
                    this.execute(child, indent, dataView)
                }
                break;
            case ASTNodeType.Multiplicative:
            case ASTNodeType.Additive: {
                lvalue = this.execute(node.children[0], indent, dataView);
                if (lvalue) emit(dataView, lvalue);
                rvalue = this.execute(node.children[1], indent, dataView);
                if (rvalue) emit(dataView, rvalue);
                if (node.text == "+") emit1(OpType.ADD, dataView);
                else if (node.text == "-") emit1(OpType.MINUS, dataView);
                else if (node.text == "*") emit1(OpType.MUTI, dataView);
                else if (node.text == "/") emit1(OpType.DIVID, dataView);
                break;
            }
            case ASTNodeType.IntLiteral:
                return parseInt(node.text);
            case ASTNodeType.Identifier:
                return node.text;
            case ASTNodeType.AssignmentStmt:
            case ASTNodeType.IntDeclaration:
                value = this.execute(node.children[0], indent, dataView);
                if (value) emit(dataView, value);
                emit2(node.type == ASTNodeType.AssignmentStmt ? OpType.ASSIGN : OpType.INT_DECL, dataView, node.text);
                break;
        }
        function emit(dataView: DataView, value: any) {
            emit2(typeof value == "string" ? OpType.NAME : OpType.NUMBER, dataView, value);
        }
        function emit1(op: OpType, dataView: DataView) {
            that.emit1(op, dataView);
            // console.log(indent, node.type, node.text, "emit1", to_optype_str(op), that.offset);
        }
        function emit2(op: OpType, dataView: DataView, val: string | number) {
            that.emit2(op, dataView, val);
            // console.log(indent, node.type, node.text, "emit2", to_optype_str(op), val, that.offset);
        }
    }
    emit1(op: OpType, dataView: DataView) {
        dataView.setInt8(this.offset, op);
        this.offset += 8;
    }
    emit2(op: OpType, dataView: DataView, val: string | number) {
        dataView.setInt8(this.offset, op);
        this.offset += 8;
        if (typeof val == "string") {
            let strUint8Array = str2Uint8Array(val);
            dataView.setInt32(this.offset, strUint8Array.byteLength);
            this.offset += 32;
            const baseUnit8Array = new Uint8Array(dataView.buffer, this.offset, strUint8Array.byteLength);
            baseUnit8Array.set(strUint8Array);
            this.offset += strUint8Array.byteLength;
        } else if (typeof val == "number") {
            dataView.setInt32(this.offset, val);
            this.offset += 32;
        }
    }
}